File: game/scripts/vscripts/abilities/heroes/invoker/invoker_extra/invoker_extra.lua
invoker_extra = class({})

--[[
    amounts = {
        wex = x, R
        quas = y, Q
        exort = z, E
    }
]]

local invoker_abilities = {
    ['120'] = "invoker_walk_custom",
	['012'] = "invoker_spirit_custom",
	['021'] = "invoker_wall_custom",
    ['003'] = "invoker_sun_custom",
	['030'] = "invoker_snap_custom",
	['300'] = "invoker_emp_custom",
	['111'] = "invoker_blast_custom",
	['201'] = "invoker_alacrity_custom",
    ['210'] = "invoker_tornado_custom",
    ['102'] = "invoker_meteor_custom",
}

function invoker_extra:OnSpellStart()
    local caster = self:GetCaster()
    local orb_amounts = caster:FindModifierByName("modifier_invoker_helper"):GetOrbAmounts()
    local ability_code = tostring(orb_amounts['wex'] .. orb_amounts['quas'] .. orb_amounts['exort'])

    if invoker_abilities[ability_code] then
        local incoming_ability_name = invoker_abilities[ability_code]
        local incoming_ability = caster:FindAbilityByName(incoming_ability_name)
        local first_ability = caster:GetAbilityByIndex(7)
        local second_ability = caster:GetAbilityByIndex(8)
        local outcoming_ability = nil
        local enable_outcoming_ability = false

        if first_ability == incoming_ability then
            outcoming_ability = second_ability
            enable_outcoming_ability = true
        elseif second_ability == incoming_ability then
            outcoming_ability = first_ability
            enable_outcoming_ability = true
        elseif first_ability:GetName() == "invoker_empty1" or first_ability:GetName() == "invoker_empty2" then
            outcoming_ability = first_ability
        elseif second_ability:GetName() == "invoker_empty1" or second_ability:GetName() == "invoker_empty2" then
            outcoming_ability = second_ability
        else
            caster:SwapAbilities(
                first_ability:GetName(), 
                second_ability:GetName(), 
                true,
                true
            )
            outcoming_ability = caster:GetAbilityByIndex(7)
        end

        caster:SwapAbilities(
            incoming_ability_name, 
            outcoming_ability:GetName(), 
            true,
            enable_outcoming_ability
        )
        EmitSoundOn("Hero_Invoker.Invoke", caster)
    end
end

File: src/vscripts/abilities/heroes/invoker/invoker_extra.ts
import { registerAbility } from "../../../lib/dota_ts_adapter";
import { CustomAbility } from "../../framework/custom_ability";
import { InvokerOrbType } from "./invoker_common";
import { ModifierInvokerHelper } from "./invoker_helper";

type ComboDefinition = [[InvokerOrbType, InvokerOrbType, InvokerOrbType], string];

const comboDefinitions: ComboDefinition[] = [
    [["q", "q", "q"], "invoker_snap_custom"],
    [["q", "q", "w"], "invoker_walk_custom"],
    [["q", "q", "e"], "invoker_wall_custom"],

    [["w", "w", "w"], "invoker_emp_custom"],
    [["w", "w", "q"], "invoker_tornado_custom"],
    [["w", "w", "e"], "invoker_alacrity_custom"],

    [["e", "e", "e"], "invoker_sun_custom"],
    [["e", "e", "q"], "invoker_spirit_custom"],
    [["e", "e", "w"], "invoker_meteor_custom"],

    [["q", "w", "e"], "invoker_blast_custom"]
];

type Combo = {
    orbs: Record<InvokerOrbType, number>;
    abilityName: string;
};

const combos: Combo[] = comboDefinitions.map(([combo, ability]) => {
    return {
        orbs: serializeOrbs(combo),
        abilityName: ability
    };
});

function serializeOrbs(orbs: InvokerOrbType[]) {
    function count(orb: InvokerOrbType) {
        return orbs.filter(x => x == orb).length;
    }

    return {
        q: count("q"),
        w: count("w"),
        e: count("e")
    };
}

@registerAbility("invoker_extra")
export class InvokerExtra extends CustomAbility {
    OnSpellStart() {
        const orbs = serializeOrbs(ModifierInvokerHelper.findOne(this.caster)?.orbs ?? []);
        const combo = combos.find(combo => combo.orbs.q == orbs.q && combo.orbs.w == orbs.w && combo.orbs.e == orbs.e);

        if (combo) {
            const invokedAbility = this.caster.FindAbilityByName(combo.abilityName);
            const currentAbilities = [this.caster.GetAbilityByIndex(7), this.caster.GetAbilityByIndex(8)];

            if (!invokedAbility || !currentAbilities[0] || !currentAbilities[1]) {
                return;
            }

            this.caster.SwapAbilities(currentAbilities[0].GetName(), currentAbilities[1].GetName(), true, true);

            if (currentAbilities.includes(invokedAbility)) {
                return;
            }

            this.caster.SwapAbilities(invokedAbility.GetName(), currentAbilities[0].GetName(), true, false);
            EmitSoundOn("Hero_Invoker.Invoke", this.caster);
        }
    }
}
